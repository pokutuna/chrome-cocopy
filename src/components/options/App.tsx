import {ThemeProvider} from 'styled-components';
import {Routes, Route} from 'react-router-dom';

import {theme} from '../common/Theme';
import {MainColumn, Title, VersionLabel} from './Parts';
import {FunctionList} from './FunctionList';
import {Links} from './Links';
import {Hint, DebuggingHint} from './Hints';
import {InstallFunction} from './InstallFunction';

const PageRoot = () => (
  <>
    <FunctionList />
    <Hint />
    <DebuggingHint />
    <Links />
  </>
);

const PageInstall = () => (
  <>
    <InstallFunction />
    <Hint />
    <Links />
  </>
);

export const App = () => {
  return (
    <ThemeProvider theme={theme}>
      <MainColumn>
        <Title />
        <Routes>
          <Route path="/" element={<PageRoot />} />
          <Route path="/install" element={<PageInstall />} />
        </Routes>
        <VersionLabel />
      </MainColumn>
    </ThemeProvider>
  );
};
