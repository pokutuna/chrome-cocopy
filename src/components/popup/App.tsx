import React from 'react';
import {ThemeProvider} from 'styled-components';

import {theme} from '../common/Theme';
import {PopupHeader} from './Header';
import {FunctionList} from './FunctionList';

export const App = () => {
  return (
    <ThemeProvider theme={theme}>
      <PopupHeader />
      <FunctionList />
    </ThemeProvider>
  );
};
