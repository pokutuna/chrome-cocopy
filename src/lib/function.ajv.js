"use strict";module.exports = validate20;module.exports.default = validate20;const schema22 = {"type":"object","properties":{"id":{"type":"string"},"name":{"type":"string"},"code":{"type":"string"},"pattern":{"type":"string"},"theme":{"type":"object","properties":{"textColor":{"type":"string","pattern":"^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$"},"backgroundColor":{"type":"string","pattern":"^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$"}},"required":["textColor","backgroundColor"],"additionalProperties":false},"version":{"const":1}},"required":["id","name","code","theme","version"],"additionalProperties":false};const pattern0 = new RegExp("^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$", "u");const func0 = require("ajv/dist/runtime/equal").default;function validate20(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){let missing0;if((((((data.id === undefined) && (missing0 = "id")) || ((data.name === undefined) && (missing0 = "name"))) || ((data.code === undefined) && (missing0 = "code"))) || ((data.theme === undefined) && (missing0 = "theme"))) || ((data.version === undefined) && (missing0 = "version"))){validate20.errors = [{instancePath,schemaPath:"#/required",keyword:"required",params:{missingProperty: missing0},message:"must have required property '"+missing0+"'"}];return false;}else {const _errs1 = errors;for(const key0 in data){if(!((((((key0 === "id") || (key0 === "name")) || (key0 === "code")) || (key0 === "pattern")) || (key0 === "theme")) || (key0 === "version"))){validate20.errors = [{instancePath,schemaPath:"#/additionalProperties",keyword:"additionalProperties",params:{additionalProperty: key0},message:"must NOT have additional properties"}];return false;break;}}if(_errs1 === errors){if(data.id !== undefined){const _errs2 = errors;if(typeof data.id !== "string"){validate20.errors = [{instancePath:instancePath+"/id",schemaPath:"#/properties/id/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs2 === errors;}else {var valid0 = true;}if(valid0){if(data.name !== undefined){const _errs4 = errors;if(typeof data.name !== "string"){validate20.errors = [{instancePath:instancePath+"/name",schemaPath:"#/properties/name/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs4 === errors;}else {var valid0 = true;}if(valid0){if(data.code !== undefined){const _errs6 = errors;if(typeof data.code !== "string"){validate20.errors = [{instancePath:instancePath+"/code",schemaPath:"#/properties/code/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs6 === errors;}else {var valid0 = true;}if(valid0){if(data.pattern !== undefined){const _errs8 = errors;if(typeof data.pattern !== "string"){validate20.errors = [{instancePath:instancePath+"/pattern",schemaPath:"#/properties/pattern/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs8 === errors;}else {var valid0 = true;}if(valid0){if(data.theme !== undefined){let data4 = data.theme;const _errs10 = errors;if(errors === _errs10){if(data4 && typeof data4 == "object" && !Array.isArray(data4)){let missing1;if(((data4.textColor === undefined) && (missing1 = "textColor")) || ((data4.backgroundColor === undefined) && (missing1 = "backgroundColor"))){validate20.errors = [{instancePath:instancePath+"/theme",schemaPath:"#/properties/theme/required",keyword:"required",params:{missingProperty: missing1},message:"must have required property '"+missing1+"'"}];return false;}else {const _errs12 = errors;for(const key1 in data4){if(!((key1 === "textColor") || (key1 === "backgroundColor"))){validate20.errors = [{instancePath:instancePath+"/theme",schemaPath:"#/properties/theme/additionalProperties",keyword:"additionalProperties",params:{additionalProperty: key1},message:"must NOT have additional properties"}];return false;break;}}if(_errs12 === errors){if(data4.textColor !== undefined){let data5 = data4.textColor;const _errs13 = errors;if(errors === _errs13){if(typeof data5 === "string"){if(!pattern0.test(data5)){validate20.errors = [{instancePath:instancePath+"/theme/textColor",schemaPath:"#/properties/theme/properties/textColor/pattern",keyword:"pattern",params:{pattern: "^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$"},message:"must match pattern \""+"^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$"+"\""}];return false;}}else {validate20.errors = [{instancePath:instancePath+"/theme/textColor",schemaPath:"#/properties/theme/properties/textColor/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}}var valid1 = _errs13 === errors;}else {var valid1 = true;}if(valid1){if(data4.backgroundColor !== undefined){let data6 = data4.backgroundColor;const _errs15 = errors;if(errors === _errs15){if(typeof data6 === "string"){if(!pattern0.test(data6)){validate20.errors = [{instancePath:instancePath+"/theme/backgroundColor",schemaPath:"#/properties/theme/properties/backgroundColor/pattern",keyword:"pattern",params:{pattern: "^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$"},message:"must match pattern \""+"^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$"+"\""}];return false;}}else {validate20.errors = [{instancePath:instancePath+"/theme/backgroundColor",schemaPath:"#/properties/theme/properties/backgroundColor/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}}var valid1 = _errs15 === errors;}else {var valid1 = true;}}}}}else {validate20.errors = [{instancePath:instancePath+"/theme",schemaPath:"#/properties/theme/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}var valid0 = _errs10 === errors;}else {var valid0 = true;}if(valid0){if(data.version !== undefined){const _errs17 = errors;if(!func0(data.version, 1)){validate20.errors = [{instancePath:instancePath+"/version",schemaPath:"#/properties/version/const",keyword:"const",params:{allowedValue: 1},message:"must be equal to constant"}];return false;}var valid0 = _errs17 === errors;}else {var valid0 = true;}}}}}}}}}else {validate20.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate20.errors = vErrors;return errors === 0;}