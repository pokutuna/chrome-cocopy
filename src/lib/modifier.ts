export type Modifier = {
  alt: boolean;
  ctrl: boolean;
  meta: boolean;
  shift: boolean;
};

export const EMPTY_MODIFIER: Modifier = {
  alt: false,
  ctrl: false,
  meta: false,
  shift: false,
};

export function isModifier(input: any): input is Modifier {
  return (
    typeof input === 'object' &&
    typeof input.alt === 'boolean' &&
    typeof input.ctrl === 'boolean' &&
    typeof input.meta === 'boolean' &&
    typeof input.shift === 'boolean'
  );
}

export function getModifier(e: KeyboardEvent): Modifier {
  return {
    alt: e.altKey,
    ctrl: e.ctrlKey,
    meta: e.metaKey,
    shift: e.shiftKey,
  };
}

export function hasModifierChanged(prev: Modifier, next: Modifier): boolean {
  return (
    prev.alt !== next.alt ||
    prev.ctrl !== next.ctrl ||
    prev.meta !== next.meta ||
    prev.shift !== next.shift
  );
}
