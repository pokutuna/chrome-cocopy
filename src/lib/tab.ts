export async function getActiveTab(): Promise<chrome.tabs.Tab> {
  const tabs = await chrome.tabs.query({active: true, currentWindow: true});
  if (tabs && tabs[0]) return tabs[0];
  throw new Error('no active tabs');
}

function timeout(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Execute short code as a content_script with timeout
 * @param tab
 * @param code js string returns a string
 */
async function executeInTab(
  tab: chrome.tabs.Tab,
  fn: () => void
): Promise<string | undefined> {
  const p = new Promise<string>(resolve => {
    return chrome.scripting.executeScript(
      {target: {tabId: tab.id!}, func: fn},
      results => {
        resolve(results?.[0].result);
      }
    );
  });

  const result = await Promise.race([p, timeout(2000)]);
  return typeof result === 'string' ? result : undefined;
}

export async function getTabContent(
  tab: chrome.tabs.Tab
): Promise<string | undefined> {
  return executeInTab(tab, _getTabContent);
}

function _getTabContent(): string {
  return document.documentElement.outerHTML;
}

export async function getSelectionText(
  tab: chrome.tabs.Tab
): Promise<string | undefined> {
  return executeInTab(tab, _getSelectionText);
}

function _getSelectionText(): string {
  return (window.getSelection() || '').toString();
}
