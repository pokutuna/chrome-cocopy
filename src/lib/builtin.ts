import {CopyFunction} from './function';

export const initialCode = `
/**
 * Return value will be copied to clipboard.
 * @param {Object} page
 * @returns {(string|undefined|Promise)}
 */
({title, url, content, selectingText, modifier}) => {
  return [title, url].join(' ');
}
`.trim();

const copyAsMarkdownFn = `
/**
 * Copy link as Markdown.
 */
({title, url}) => {
  const escaped = title.replace(/[\\[\\]]/g, '\\\\$&');
  return \`[\${escaped}](\${url})\`;
};
`.trim();

export const copyAsMarkdown: CopyFunction = {
  id: 'builtin-markdown',
  name: 'Markdown: [title](url)',
  code: copyAsMarkdownFn,
  pattern: undefined,
  version: 1,
  theme: {
    textColor: '#000000',
    backgroundColor: '#f5f5f5',
  },
};

const copyAsHTMLFn = `
/**
 * Copy as anchor element.
 *
 * You can use mustache template with \`render\` function.
 * @see https://github.com/janl/mustache.js
 */
page => render('<a href="{{&url}}">{{title}}</a>', page);`.trim();

export const copyAsHTML: CopyFunction = {
  id: 'builtin-html',
  name: 'HTML: <a href={url}>{title}</a>',
  code: copyAsHTMLFn,
  pattern: undefined,
  version: 1,
  theme: {
    textColor: '#FFFFFF',
    backgroundColor: '#ff5722',
  },
};

export const defaultFunctions: CopyFunction[] = [copyAsMarkdown, copyAsHTML];
