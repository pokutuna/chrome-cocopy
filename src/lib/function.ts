import {Page} from './page';
import {Library} from './library';
import {initialCode} from './builtin';
import {textColorFromBgColor} from './util';
import validate from './function.ajv';

export const currentVersion = 1;

export interface RichContent {
  html: string;
  text: string;
}

export type CopyResult = string | RichContent | null | undefined;
export type CopyFn = (
  this: Library,
  t: Page
) => CopyResult | Promise<CopyResult>;

export interface CopyFunction {
  id: string;
  name: string;
  code: string;
  pattern?: string;
  version: number;
  theme: {
    textColor: string;
    backgroundColor: string;
  };
}

export function generateId(): string {
  return `function-${new Date().getTime()}`;
}

export function newFunction(): CopyFunction {
  const id = generateId();
  const backgroundColor =
    colorPalette[Math.floor(Math.random() * colorPalette.length)];
  const textColor = textColorFromBgColor(backgroundColor);

  return {
    id,
    name: 'function name',
    code: initialCode,
    pattern: '',
    theme: {
      textColor,
      backgroundColor,
    },
    version: currentVersion,
  };
}

export function filterFunctions<T extends CopyFunction>(
  url: string,
  functions: T[]
): T[] {
  return functions.filter(f =>
    f.pattern ? new RegExp(f.pattern).test(url) : true
  );
}

export const colorPalette = [
  '#F44336',
  '#E91E63',
  '#9C27B0',
  '#673AB7',
  '#3F51B5',
  '#2196F3',
  '#03A9F4',
  '#00BCD4',
  '#009688',
  '#4CAF50',
  '#8BC34A',
  '#CDDC39',
  '#FFEB3B',
  '#FFC107',
  '#FFC107',
  '#FF5722',
  '#795548',
  '#9E9E9E',
  '#607D8B',
];

export function isCopyFunction(f: any): f is CopyFunction {
  return validate(f) as boolean;
}
