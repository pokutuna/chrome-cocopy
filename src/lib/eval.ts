import {Page, isPage} from './page';
import {CopyResult, RichContent} from './function';
import {Modifier, isModifier} from './modifier';

export type FunctionArgument = Page & {
  modifier: Modifier;
};

export function isFunctionArgument(input: any): input is FunctionArgument {
  return isPage(input) && isModifier((input as any).modifier);
}

type FunctionEvalPayload = {
  command: 'eval';
  code: string;
  arg: FunctionArgument;
};

type ParseEvalPayload = {
  command: 'parse';
  code: string;
};

export type EvalPayload = FunctionEvalPayload | ParseEvalPayload;

export type EvalError = {
  type: 'ParseError' | 'ExecutionError' | 'ReturnsEmpty';
  name: string;
  message: string;
};

export interface EvalResult {
  result: CopyResult;
  error?: EvalError;
}

export function isEvalPayload(input: any): input is EvalPayload {
  return (
    typeof input.code === 'string' &&
    (input.command === 'parse' ||
      (input.command === 'eval' && isFunctionArgument(input.arg)))
  );
}

function isAcceptableResult(input: any): input is CopyResult {
  return (
    typeof input === 'string' ||
    typeof input === 'number' ||
    typeof input === 'undefined' ||
    input === null ||
    isRichContent(input)
  );
}

export function isRichContent(input: any): input is RichContent {
  return typeof input === 'object' && 'html' in input && 'text' in input;
}

function isEmpty(input: any): boolean {
  return typeof input === 'undefined' || input === null;
}

function error(type: EvalError['type'], err: Error): EvalError {
  return {
    type,
    name: err.name,
    message: err.message,
  };
}

export async function evaluate(request: EvalPayload): Promise<EvalResult> {
  if (request.command !== 'parse' && request.command !== 'eval') {
    throw Error('unexpected command');
  }

  // parse
  let fn: Function;
  try {
    fn = eval.call(undefined, request.code);
    if (typeof fn !== 'function') {
      throw new Error('code is not a function');
    }
  } catch (e) {
    return {
      result: null,
      error: error(
        'ParseError',
        e instanceof Error ? e : new Error(`Unknown error: ${e}`)
      ),
    };
  }

  if (request.command === 'parse') {
    return {result: 'ok'};
  }

  // excute
  let result: any;
  try {
    result = await Promise.resolve(fn.call(undefined, request.arg));
    if (!isAcceptableResult(result)) {
      throw new Error(
        'returning value is not one of (string | number | { html: string, text: string } | null | undefined)'
      );
    }
  } catch (e) {
    return {
      result,
      error: error(
        'ExecutionError',
        e instanceof Error ? e : new Error(`Unknown error: ${e}`)
      ),
    };
  }

  if (isEmpty(result)) {
    return {
      result,
      error: error(
        'ReturnsEmpty',
        new Error('returning value is empty (undefined or null)')
      ),
    };
  }

  return {result};
}
