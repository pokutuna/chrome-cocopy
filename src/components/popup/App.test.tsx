import {h} from 'preact';

import {render, screen, waitFor} from '@testing-library/preact';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import {chrome} from 'jest-chrome';

import {defaultFunctions} from '../../lib/builtin';
import {App} from './App';

test('render popup', async () => {
  chrome.tabs.query.mockImplementation(
    async () => [{url: 'https://example.test/page'}] as chrome.tabs.Tab[]
  );
  chrome.storage.sync.get.mockImplementation((_, cb) =>
    cb({functions: defaultFunctions})
  );
  chrome.runtime.getManifest.mockImplementation(
    () => ({version_name: 'Build v0.0.0'} as chrome.runtime.Manifest)
  );

  render(<App />);

  await waitFor(() =>
    expect(screen.getByText(defaultFunctions[0].name)).toBeInTheDocument()
  );

  expect(document.body).toMatchSnapshot();
});
