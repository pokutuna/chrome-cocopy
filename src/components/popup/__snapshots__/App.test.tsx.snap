// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render popup 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c1 {
  width: 84px;
  height: 40px;
  background: no-repeat center/contain url('img/logo.png');
  margin: 0.25rem;
  padding: 0 0.25rem;
  -webkit-filter: none;
  filter: none;
}

.c2 {
  position: absolute;
  right: 0;
  font-size: 1.25rem;
  margin-right: 1rem;
  color: #666;
}

.c2:hover,
.c2:focus {
  color: #333;
  -webkit-animation: cilQsd 1.5s linear infinite;
  animation: cilQsd 1.5s linear infinite;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 380px;
  height: 32px;
  padding: 0.25rem 0.5rem;
  color: #000000;
  background-color: #f5f5f5;
  cursor: pointer;
}

.c3:focus,
.c3:hover {
  -webkit-filter: brightness(110%);
  filter: brightness(110%);
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 380px;
  height: 32px;
  padding: 0.25rem 0.5rem;
  color: #FFFFFF;
  background-color: #ff5722;
  cursor: pointer;
}

.c8:focus,
.c8:hover {
  -webkit-filter: brightness(110%);
  filter: brightness(110%);
}

.c5 {
  width: 32px;
}

.c6 {
  text-align: center;
  padding: 0 0.25rem;
  margin-right: 0.5rem;
  font-size: 1.125rem;
  font-family: Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;
  border-radius: 0.25rem;
  border: 2px solid #000000;
  box-shadow: 2px 2px 0px 0px rgba(0,0,0,0.3);
}

.c10 {
  text-align: center;
  padding: 0 0.25rem;
  margin-right: 0.5rem;
  font-size: 1.125rem;
  font-family: Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;
  border-radius: 0.25rem;
  border: 2px solid #FFFFFF;
  box-shadow: 2px 2px 0px 0px rgba(0,0,0,0.3);
}

.c7 {
  font-size: 1rem;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  -webkit-animation-name: none !important;
  animation-name: none !important;
  background-color: #f5f5f5;
  background-image: none;
  background-position: 0% !important;
}

.c9 {
  -webkit-animation-name: none !important;
  animation-name: none !important;
  background-color: #ff5722;
  background-image: none;
  background-position: 0% !important;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
      />
      <a
        class="c2"
        href="/options.html"
        target="_blank"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-gear "
          data-icon="gear"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M495.9 166.6C499.2 175.2 496.4 184.9 489.6 191.2L446.3 230.6C447.4 238.9 448 247.4 448 256C448 264.6 447.4 273.1 446.3 281.4L489.6 320.8C496.4 327.1 499.2 336.8 495.9 345.4C491.5 357.3 486.2 368.8 480.2 379.7L475.5 387.8C468.9 398.8 461.5 409.2 453.4 419.1C447.4 426.2 437.7 428.7 428.9 425.9L373.2 408.1C359.8 418.4 344.1 427 329.2 433.6L316.7 490.7C314.7 499.7 307.7 506.1 298.5 508.5C284.7 510.8 270.5 512 255.1 512C241.5 512 227.3 510.8 213.5 508.5C204.3 506.1 197.3 499.7 195.3 490.7L182.8 433.6C167 427 152.2 418.4 138.8 408.1L83.14 425.9C74.3 428.7 64.55 426.2 58.63 419.1C50.52 409.2 43.12 398.8 36.52 387.8L31.84 379.7C25.77 368.8 20.49 357.3 16.06 345.4C12.82 336.8 15.55 327.1 22.41 320.8L65.67 281.4C64.57 273.1 64 264.6 64 256C64 247.4 64.57 238.9 65.67 230.6L22.41 191.2C15.55 184.9 12.82 175.3 16.06 166.6C20.49 154.7 25.78 143.2 31.84 132.3L36.51 124.2C43.12 113.2 50.52 102.8 58.63 92.95C64.55 85.8 74.3 83.32 83.14 86.14L138.8 103.9C152.2 93.56 167 84.96 182.8 78.43L195.3 21.33C197.3 12.25 204.3 5.04 213.5 3.51C227.3 1.201 241.5 0 256 0C270.5 0 284.7 1.201 298.5 3.51C307.7 5.04 314.7 12.25 316.7 21.33L329.2 78.43C344.1 84.96 359.8 93.56 373.2 103.9L428.9 86.14C437.7 83.32 447.4 85.8 453.4 92.95C461.5 102.8 468.9 113.2 475.5 124.2L480.2 132.3C486.2 143.2 491.5 154.7 495.9 166.6V166.6zM256 336C300.2 336 336 300.2 336 255.1C336 211.8 300.2 175.1 256 175.1C211.8 175.1 176 211.8 176 255.1C176 300.2 211.8 336 256 336z"
            fill="currentColor"
          />
        </svg>
      </a>
    </div>
    <div
      class="c3 c4"
      tabindex="1"
    >
      <div
        class="c5"
      >
        <kbd
          class="c6"
        >
          1
        </kbd>
      </div>
      <div
        class="c7"
      >
        Markdown: [title](url)
      </div>
    </div>
    <div
      class="c8 c9"
      tabindex="1"
    >
      <div
        class="c5"
      >
        <kbd
          class="c10"
        >
          2
        </kbd>
      </div>
      <div
        class="c7"
      >
        HTML: &lt;a href={url}&gt;{title}&lt;/a&gt;
      </div>
    </div>
  </div>
</body>
`;
